File: rd-net/Test.RdFramework/RdIdHeavyCollisionTest.cs
using System;
using System.Collections.Generic;
using JetBrains.Collections.Viewable;
using JetBrains.Lifetimes;
using JetBrains.Rd;
using JetBrains.Rd.Base;
using JetBrains.Rd.Impl;
using JetBrains.Serialization;
using NUnit.Framework;

namespace Test.RdFramework
{
  [TestFixture]
  public class RdIdHeavyCollisionTest
  {
    private const int NodeCount = 20000;
    private const int Seed = 1;

    private static IIdentities Legacy()
    {
#pragma warning disable CS0618 // Type or member is obsolete
      return new Identities(IdKind.Server);
#pragma warning restore CS0618
    }

    private static IIdentities Sequential() => new SequentialIdentities(IdKind.Server);

    private static IEnumerable<TestCaseData> Cases()
    {
      yield return new TestCaseData((Func<IIdentities>)Legacy, true)
        .SetName("legacy Next COLLIDES in a large decoupled graph");
      yield return new TestCaseData((Func<IIdentities>)Sequential, false)
        .SetName("SequentialIdentities stays unique across a large decoupled graph");
    }

    [TestCaseSource(nameof(Cases))]
    public void HeavyDecoupledGraph(Func<IIdentities> identitiesFactory, bool expectCollision)
    {
      var ld = new LifetimeDefinition();
      try
      {
        var lifetime = ld.Lifetime;
        var protocol = new Protocol("Heavy", new Serializers(), identitiesFactory(),
          SynchronousScheduler.Instance, new StubWire(), lifetime);
        var identities = protocol.Identities;

        var root = new Node();
        root.Identify(identities, new RdId(1), stable: false);
        var nodes = new List<Node> { root };

        using (AllowBindCookie.Create())
        {
          root.PreBind(lifetime, protocol, "root");

          var random = new Random(Seed);
          TestDelegate build = () =>
          {
            for (var k = 0; k < NodeCount; k++)
            {
              // Attach to an arbitrary earlier node: decoupling parent id from allocation order is what
              // exposes legacy Next's poor distribution (a balanced tree would not collide).
              var parent = nodes[random.Next(nodes.Count)];
              var child = new Node();
              child.Identify(identities, identities.Next(parent.RdId), stable: false);
              child.PreBind(lifetime, parent, "n" + k); // registers; throws on a duplicate RdId
              nodes.Add(child);
            }
          };

          if (expectCollision)
            Assert.Throws<ArgumentException>(build, "legacy dynamic ids must collide in a large decoupled graph");
          else
            Assert.DoesNotThrow(build, "SequentialIdentities must stay unique across the whole graph");
        }
      }
      finally
      {
        ld.Terminate();
      }
    }

    private sealed class Node : RdBindableBase
    {
    }

    private sealed class StubWire : IWire
    {
      public bool IsStub => true;
      public void Send<TParam>(RdId id, TParam param, Action<TParam, UnsafeWriter> writer) { }
      public void Advise(Lifetime lifetime, IRdWireable entity) { }
      public ProtocolContexts Contexts { get; set; } = null!;
      public IRdWireable? TryGetById(RdId rdId) => null;
    }
  }
}

File: rd-kt/rd-framework/src/test/kotlin/com/jetbrains/rd/framework/test/cases/RdIdHeavyCollisionTest.kt
package com.jetbrains.rd.framework.test.cases

import com.jetbrains.rd.framework.*
import com.jetbrains.rd.framework.base.AllowBindingCookie
import com.jetbrains.rd.framework.base.RdBindableBase
import com.jetbrains.rd.framework.test.util.TestScheduler
import com.jetbrains.rd.framework.test.util.TestWire
import com.jetbrains.rd.util.lifetime.LifetimeDefinition
import kotlin.test.Test
import kotlin.test.assertFailsWith

class RdIdHeavyCollisionTest {

    private companion object {
        const val NODE_COUNT = 20000
        const val SEED = 1L
    }

    private class Node : RdBindableBase()

    @Suppress("DEPRECATION")
    private fun legacy(): IIdentities = Identities(IdKind.Server)
    private fun sequential(): IIdentities = SequentialIdentities(IdKind.Server)

    @Test
    fun `legacy dynamic ids collide in a large decoupled graph`() {
        assertFailsWith<IllegalArgumentException> { buildHeavyGraph(legacy()) }
    }

    @Test
    fun `SequentialIdentities stays unique across a large decoupled graph`() {
        buildHeavyGraph(sequential()) // must complete without a duplicate-id exception
    }

    @Suppress("DEPRECATION")
    private fun buildHeavyGraph(identities: IIdentities) {
        val ld = LifetimeDefinition()
        try {
            val lifetime = ld.lifetime
            val protocol = Protocol("Heavy", Serializers(), identities, TestScheduler, TestWire(TestScheduler), lifetime)

            val root = Node().also { it.identify(identities, RdId(1L), false) }
            val nodes = ArrayList<Node>().apply { add(root) }

            AllowBindingCookie.allowBind {
                root.preBind(lifetime, protocol, "root")

                val random = java.util.Random(SEED)
                for (k in 0 until NODE_COUNT) {
                    // Attach to an arbitrary earlier node: decoupling parent id from allocation order is
                    // what exposes legacy next's poor distribution (a balanced tree would not collide).
                    val parent = nodes[random.nextInt(nodes.size)]
                    val child = Node()
                    child.identify(identities, identities.next(parent.rdid), false)
                    child.preBind(lifetime, parent, "n$k") // registers; throws on a duplicate RdId
                    nodes.add(child)
                }
            }
        } finally {
            ld.terminate()
        }
    }
}
